### PR TITLE
MAINT: stats.sampling.NumericalInverseHermite: private distribution methods need at least 1d arrays

### LINKED ISSUE #17155: BUG: stats: Bug in XSLOW tests in TestNumericalInverseHermite
### Describe your issue.

When the environment variable `SCIPY_XSLOW` is set, I get two failures of a test in the `TestNumericalInverseHermite` test class.  I am also use a development version of numpy (version `'1.24.0.dev0+837.g9c5a2b41c'`).

Here's the transcript of running just the `TestNumericalInverseHermite` tests:

```
$ SCIPY_XSLOW=1 python dev.py test -m full -j 8 -t scipy.stats.tests.test_sampling::TestNumericalInverseHermite
💻  ninja -C /home/warren/repos/git/forks/scipy/build
ninja: Entering directory `/home/warren/repos/git/forks/scipy/build'
[2/2] Generating scipy/generate-config with a custom command
Build OK
💻  meson install -C build --only-changed
Installing, see meson-install.log...
Installation OK
SciPy from development installed path at: /home/warren/repos/git/forks/scipy/build-install/lib/python3.10/site-packages
Running tests for scipy version:1.10.0.dev0+2006.527f2dd, installed at:/home/warren/repos/git/forks/scipy/build-install/lib/python3.10/site-packages/scipy
============================================================== test session starts ===============================================================
platform linux -- Python 3.10.1, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/warren/repos/git/forks/scipy, configfile: pytest.ini
plugins: forked-1.4.0, timeout-2.1.0, hypothesis-6.35.0, xdist-2.5.0
gw0 [145] / gw1 [145] / gw2 [145] / gw3 [145] / gw4 [145] / gw5 [145] / gw6 [145] / gw7 [145]
...............x............x.......x.........xxx....ss.x...x..............x.x...........xx.xs.s..........x.........................F....x [ 95%]
..F....                                                                                                                                    [100%]
==================================================================== FAILURES ====================================================================
__________________________________ TestNumericalInverseHermite.test_basic_all_scipy_dists[truncnorm-shapes101] ___________________________________
[gw7] linux -- Python 3.10.1 /home/warren/py3.10.1/bin/python
scipy/stats/tests/test_sampling.py:1019: in test_basic_all_scipy_dists
    fni = NumericalInverseHermite(dist)
        dist       = <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7f9e0d8fbc70>
        distname   = 'truncnorm'
        fail_dists = {'beta', 'gausshyper', 'genhyperbolic', 'geninvgauss', 'invgauss', 'kappa4', ...}
        self       = <scipy.stats.tests.test_sampling.TestNumericalInverseHermite object at 0x7f9e0d8fb1c0>
        shapes     = (0.1, 2.0)
        slow_dists = {'ksone', 'kstwo', 'levy_stable', 'skewnorm'}
        sup        = <numpy.testing._private.utils.suppress_warnings object at 0x7f9e0d8fbaf0>
scipy/stats/unuran_wrapper.pyx:2020: in scipy.stats.unuran_wrapper.NumericalInverseHermite.__cinit__
    ???
scipy/stats/unuran_wrapper.pyx:1996: in scipy.stats.unuran_wrapper.NumericalInverseHermite.__cinit__._callback_wrapper
    ???
scipy/stats/unuran_wrapper.pyx:271: in scipy.stats.unuran_wrapper._unpack_dist.wrap_dist.cdf
    ???
scipy/stats/_continuous_distns.py:8670: in _cdf
    return np.exp(self._logcdf(x, a, b))
        a          = 0.1
        b          = 2.0
        self       = <scipy.stats._continuous_distns.truncnorm_gen object at 0x7f9e0d8fb190>
        x          = array(2.)
scipy/stats/_continuous_distns.py:8677: in _logcdf
    logcdf[i] = np.log1p(-np.exp(self._logsf(x[i], a[i], b[i])))
E   IndexError: too many indices for array: array is 0-dimensional, but 1 were indexed
        a          = array(0.1)
        b          = array(2.)
        i          = array([ True])
        logcdf     = array([0.])
        self       = <scipy.stats._continuous_distns.truncnorm_gen object at 0x7f9e0d8fb190>
        x          = array(2.)
__________________________________ TestNumericalInverseHermite.test_basic_all_scipy_dists[truncnorm-shapes100] ___________________________________
[gw1] linux -- Python 3.10.1 /home/warren/py3.10.1/bin/python
scipy/stats/tests/test_sampling.py:1019: in test_basic_all_scipy_dists
    fni = NumericalInverseHermite(dist)
        dist       = <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7fb3dff6fd00>
        distname   = 'truncnorm'
        fail_dists = {'beta', 'gausshyper', 'genhyperbolic', 'geninvgauss', 'invgauss', 'kappa4', ...}
        self       = <scipy.stats.tests.test_sampling.TestNumericalInverseHermite object at 0x7fb3dff6f6a0>
        shapes     = (-1.0978730080013919, 2.730675410903198)
        slow_dists = {'ksone', 'kstwo', 'levy_stable', 'skewnorm'}
        sup        = <numpy.testing._private.utils.suppress_warnings object at 0x7fb3dff6dd80>
scipy/stats/unuran_wrapper.pyx:2020: in scipy.stats.unuran_wrapper.NumericalInverseHermite.__cinit__
    ???
scipy/stats/unuran_wrapper.pyx:1996: in scipy.stats.unuran_wrapper.NumericalInverseHermite.__cinit__._callback_wrapper
    ???
scipy/stats/unuran_wrapper.pyx:271: in scipy.stats.unuran_wrapper._unpack_dist.wrap_dist.cdf
    ???
scipy/stats/_continuous_distns.py:8670: in _cdf
    return np.exp(self._logcdf(x, a, b))
        a          = -1.0978730080013919
        b          = 2.730675410903198
        self       = <scipy.stats._continuous_distns.truncnorm_gen object at 0x7fb3dff6d7b0>
        x          = array(2.73067541)
scipy/stats/_continuous_distns.py:8677: in _logcdf
    logcdf[i] = np.log1p(-np.exp(self._logsf(x[i], a[i], b[i])))
E   IndexError: too many indices for array: array is 0-dimensional, but 1 were indexed
        a          = array(-1.09787301)
        b          = array(2.73067541)
        i          = array([ True])
        logcdf     = array([0.])
        self       = <scipy.stats._continuous_distns.truncnorm_gen object at 0x7fb3dff6d7b0>
        x          = array(2.73067541)
============================================================ short test summary info =============================================================
FAILED scipy/stats/tests/test_sampling.py::TestNumericalInverseHermite::test_basic_all_scipy_dists[truncnorm-shapes101] - IndexError: too many ...
FAILED scipy/stats/tests/test_sampling.py::TestNumericalInverseHermite::test_basic_all_scipy_dists[truncnorm-shapes100] - IndexError: too many ...
============================================== 2 failed, 124 passed, 4 skipped, 15 xfailed in 3.32s ==============================================
warren@pop-os:~/repos/git/forks/scipy$ 

```

### Reproducing Code Example

```python
n/a
```

### Error message

```shell
n/a
```

### SciPy/NumPy/Python version information

1.10.0.dev0+2006.527f2dd 1.24.0.dev0+837.g9c5a2b41c sys.version_info(major=3, minor=10, micro=1, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-17155

#### What does this implement/fix?
This fixes gh-17155.

#### Additional information
I'm not sure if there are other implications of this change. Based on the code like:
```python3
   if res < 0:
       return 0
```
It looks like `res` is _always_ expected to have one element. So since I'm ensuring that it will be 1d with one element, shall I extract that one element from the result?

### CODE REVIEW

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 1]
```diff
@@ -264,10 +264,10 @@ cdef dict _unpack_dist(object dist, str dist_type, list meths,
                     self.support = dist.support
                 def pdf(self, x):
                     # some distributions require array inputs.
-                    x = np.asarray((x-self.loc)/self.scale)
+                    x = np.atleast_1d((x-self.loc)/self.scale)
                     return max(0, self.dist.dist._pdf(x, *self.args)/self.scale)
```
Comment (mdhaber) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:268]:
```suggestion
                    return max(0, self.dist.dist._pdf(x, *self.args)[0]/self.scale)
```

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 2]
```diff
@@ -264,10 +264,10 @@ cdef dict _unpack_dist(object dist, str dist_type, list meths,
...
                 def cdf(self, x):
-                    x = np.asarray((x-self.loc)/self.scale)
+                    x = np.atleast_1d((x-self.loc)/self.scale)
                     res = self.dist.dist._cdf(x, *self.args)
```
Comment (mdhaber) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:271]:
```suggestion
                    res = self.dist.dist._cdf(x, *self.args)[0]
```

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 3]
```diff
@@ -284,10 +284,10 @@ cdef dict _unpack_dist(object dist, str dist_type, list meths,
                     self.support = dist.support
                 def pmf(self, x):
                     # some distributions require array inputs.
-                    x = np.asarray(x-self.loc)
+                    x = np.atleast_1d(x-self.loc)
                     return max(0, self.dist.dist._pmf(x, *self.args))
```
Comment (mdhaber) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:288]:
```suggestion
                    return max(0, self.dist.dist._pmf(x, *self.args)[0])
```

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 4]
```diff
@@ -284,10 +284,10 @@ cdef dict _unpack_dist(object dist, str dist_type, list meths,
...
                 def cdf(self, x):
-                    x = np.asarray(x-self.loc)
+                    x = np.atleast_1d(x-self.loc)
                     res = self.dist.dist._cdf(x, *self.args)
```
Comment (mdhaber) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:291]:
```suggestion
                    res = self.dist.dist._cdf(x, *self.args)[0]
```

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 5]
```diff
@@ -797,7 +797,7 @@ cdef class TransformedDensityRejection(Method):
     ...         return 1-x*x
     ...     def dpdf(self, x):
     ...         return -2*x
-    ...
```
Comment (mdhaber) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:800]: My IDE does this automatically.

**Review by mdhaber — COMMENTED** [review]
Is this needed? Code seems to work without it.

**Comment by chrisb83** [comment]
> It looks like res is always expected to have one element. 

Yes, the signature of pdf, logpdf of the distribution objects that are passed to the UNURAN methods needs to be float -> float

Do you know if trunnorm is the only distribution that does not work with 0d shapes? It is not important to support truncnorm since truncation should be done in NumericalInverseHermite via the domain parameter.

**Comment by mdhaber** [comment]
> Do you know if trunnorm is the only distribution that does not work with 0d shapes?

It is the only one that is failing that test, but I don't know if it's the only one that assumes at least 1d shapes. 

>  It is not important to support truncnorm

I see. I agree that it doesn't _need_ to be supported, but if this fix is OK, it seems easier just to support them than to do something else, like raise a meaningful error for each distribution that is not supported and test all of those.

**Comment by tirthasheshpatel** [comment]
> It looks like `res` is _always_ expected to have one element. So since I'm ensuring that it will be 1d with one element, shall I extract that one element from the result?

I don't think you need to do that. NumPy arrays with one element are converted to float implicitly.

**Comment by mdhaber** [comment]
OK, fixed merge conflicts. If tests pass, I think this is ready.

**Review by tirthasheshpatel — APPROVED** [review]
Looks good now. Thanks for resolving the conflict and thanks for the PR @mdhaber!

File: scipy/stats/tests/test_sampling.py [hunk 6]
```diff
@@ -1005,10 +1005,8 @@ def test_inf_nan_domains(self, domain, err, msg):
         with pytest.raises(err, match=msg):
             NumericalInverseHermite(StandardNormal(), domain=domain)
 
-    @pytest.mark.filterwarnings('ignore::RuntimeWarning')
-    @pytest.mark.xslow
-    @pytest.mark.parametrize(("distname", "shapes"), distcont)
-    def test_basic_all_scipy_dists(self, distname, shapes):
+
```
Comment (mdhaber) [inline scipy/stats/tests/test_sampling.py:1008]:
```suggestion
```

Comment (mdhaber) [reply scipy/stats/tests/test_sampling.py:1008]: If other tests pass, this can probably get committed with [skip ci] before merge to fix the lint issue.

Comment (tirthasheshpatel) [reply scipy/stats/tests/test_sampling.py]: Just a small change, so I applied it. Hope you don't mind!

Comment (mdhaber) [reply scipy/stats/tests/test_sampling.py]: Yup, that was the intent. Thanks!

**Pull request merged**